cheroot/ssl/builtin.py

```diff
@@ -173,22 +239,164 @@ def get_environ(self, sock):
+        )
+        context.load_cert_chain(cert, private_key)
+        context.check_hostname = False
+        context.verify_mode = ssl.CERT_NONE
+
+        # Python 3+ Unix, Python 3.5+ Windows
+        client, server = socket.socketpair()
+        try:
+            # ssl handshake is blocking and needs to happen on
+            # both ends at the same time so use a thread
+            thread = threading.Thread(
+                target=lambda: context.wrap_socket(
+                    server, do_handshake_on_connect=True,
+                    server_side=True,
+                ),
+            )
+            try:
+                thread.start()
+                return context.wrap_socket(
+                    client, do_handshake_on_connect=True,
+                    server_side=False,
+                ).getpeercert()
+            finally:
+                thread.join()
+        finally:
+            client.close()
+            server.close()
+
+    def parse_cert(self, cert, private_key):
+        """Parse a certificate."""
+        # loopback_for_cert uses socket.socketpair which was only
+        # introduced in Python 3.0 for *nix and 3.5 for Windows
+        # and requires OS support (AttributeError, OSError)
+        # it also requires a private key either in its own file
+        # or combined with the cert (SSLError)
+        with suppress(AttributeError, ssl.SSLError, OSError):
+            return self.loopback_for_cert(cert, private_key)
+
+        # KLUDGE: using an undocumented, private, test method to parse a cert
+        # unfortunately, it is the only built-in way without a connection
+        # as a private, undocumented method, it may change at any time
+        # so be tolerant of *any* possible errors it may raise
+        with suppress(Exception):
+            return ssl._ssl._test_decode_cert(cert)
+
+        return {}
+
+    def env_cert_dict(self, env_prefix, parsed_cert):
+        """Return a dict of WSGI environment variables for a certificate.
+
+        E.g. SSL_CLIENT_M_VERSION, SSL_CLIENT_M_SERIAL, etc.
+        See https://httpd.apache.org/docs/2.4/mod/mod_ssl.html#envvars.
+        """
+        if not parsed_cert:
+            return {}
+
+        env = {}
+        for cert_key, env_var in self.CERT_KEY_TO_ENV.items():
+            key = '%s_%s' % (env_prefix, env_var)
+            value = parsed_cert.get(cert_key)
+            if env_var == 'SAN':
+                env.update(self.env_san_dict(key, value))
+            elif env_var.endswith('_DN'):
+                env.update(self.env_dn_dict(key, value))
+            else:
+                env[key] = str(value)
+
+        # mod_ssl 2.1+; Python 3.2+
+        if 'notBefore' in parsed_cert:
+            remain = ssl.cert_time_to_seconds(parsed_cert['notAfter'])
+            remain -= ssl.cert_time_to_seconds(parsed_cert['notBefore'])
+            remain /= 60 * 60 * 24
+            env['%s_V_REMAIN' % (env_prefix,)] = str(int(remain))
```

Comment:
Why should this be in hours?

Reply:
it should be in days as per the Apache mod_ssl documentation:
> SSL_CLIENT_V_REMAIN | string | Number of days until client's certificate expires

i've added a comment to note the timescale